compiler/rustc_restrictions/src/mut_restriction.rs

```diff
@@ -0,0 +1,163 @@
+        ..*providers
+    };
+}
+
+#[derive(Diagnostic)]
+#[diag(restrictions_mut_of_restricted_field)]
+struct MutOfRestrictedField {
+    #[primary_span]
+    mut_span: Span,
+    #[label]
+    restriction_span: Span,
+    restriction_path: String,
+}
+
+#[derive(Diagnostic)]
+#[diag(restrictions_construction_of_ty_with_mut_restricted_field)]
+struct ConstructionOfTyWithMutRestrictedField {
+    #[primary_span]
+    construction_span: Span,
+    #[label]
+    restriction_span: Span,
+    restriction_path: String,
+    #[note]
+    note: (),
+    article: &'static str,
+    description: &'static str,
+    name: String,
+}
+
+fn mut_restriction(tcx: TyCtxt<'_>, def_id: DefId) -> Restriction {
+    tracing::debug!("mut_restriction({def_id:?})");
+
+    match tcx.resolutions(()).mut_restrictions.get(&def_id) {
+        Some(restriction) => *restriction,
+        None => span_bug!(tcx.def_span(def_id), "mut restriction not found for {def_id:?}"),
+    }
+}
+
+fn check_mut_restriction(tcx: TyCtxt<'_>, def_id: LocalDefId) {
+    tracing::debug!("check_mut_restriction({def_id:?})");
+
+    let hir_id = tcx.hir().local_def_id_to_hir_id(def_id);
+    let def_id_opt_const_param = WithOptConstParam::unknown(def_id);
+
+    // FIXME(jhpratt) Does this need to be handled somehow?
+    if matches!(tcx.hir().get(hir_id), Node::AnonConst(_)) {
+        return;
+    }
+
+    let body = &tcx.mir_built(def_id_opt_const_param).borrow();
+    let mut checker = MutRestrictionChecker { tcx, body, span: body.span };
+    checker.visit_body(body);
+}
+
+/// Obtain the restriction on ADT expressions. This occurs when an ADT field has its mutability
+/// restricted.
+// This is a query to allow the compiler to cache the output. This avoids the need to recompute the
+// same information for every ADT expression.
+fn adt_expression_restriction(tcx: TyCtxt<'_>, variant_def_id: DefId) -> Restriction {
+    let res = Res::Def(tcx.def_kind(variant_def_id), variant_def_id);
+    let variant = tcx.expect_variant_res(res);
+
+    Restriction::strictest_of(
+        variant.fields.iter().map(|field| tcx.mut_restriction(field.did)),
+        tcx,
+    )
+}
+
+struct MutRestrictionChecker<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    body: &'a Body<'tcx>,
+    span: Span,
+}
+
+impl<'tcx> Visitor<'tcx> for MutRestrictionChecker<'_, 'tcx> {
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        self.span = terminator.source_info.span;
+        self.super_terminator(terminator, location);
+    }
+
+    fn visit_statement(&mut self, statement: &Statement<'tcx>, location: Location) {
+        self.span = statement.source_info.span;
+        self.super_statement(statement, location);
+    }
+
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+        if !context.is_mutating_use() {
+            self.super_place(place, context, location);
+            return;
+        }
```

Comment:
At the risk of rightward drift, I'd consider changing this to `if context.is_mutating_use()` and putting the rest of the body in the then clause. Then we don't have to remember to edit the `self.super_place` call in two places if we ever need to change it.